tests/test_glogger.py

```diff
@@ -0,0 +1,233 @@
+                logs = handler.make_batch().logs
+                serial_nos = [json.loads(log)["serial_no"] for log in logs]
+                assert_serial_nos_ok(serial_nos)
+
+
+def test_flush_when_length_threshold_reached():
+    """Test that logs are flushed when max length threshold is reached."""
+
+    class ReqHandler(BaseHTTPRequestHandler):
+        def do_POST(self):
+            json_data = json.load(self.rfile)
+            logs = json_data["logs"]
+            assert logs, "no logs!"
+            self.send_response(200, "OK")
+            self.end_headers()
+
+    with ExitStack() as exit_stack:
+        logs_server = LogsServer(("localhost", 0), ReqHandler)
+        exit_stack.callback(logs_server.server_close)
+
+        # set the interval very high because we want flush to only happen on length trigger
+        handler = HttpBatchRequestsHandler(logs_server.authority, max_total_length=10000, flush_interval=999999.0)
+        exit_stack.callback(handler.stop)
+
+        logging.basicConfig(force=True, handlers=[handler], level=0)
+        logging.info("A" * 1000)
+        logging.info("B" * 2000)
+        logging.info("C" * 3000)
+        logging.info("D" * 4000)
+        logs_server.handle_request()
+        assert logs_server.processed > 0
+
+
+def test_multiple_threads():
+    """Test that multiple threads writing simultaneously do not corrupt the buffer."""
+
+    class MockSession:
+        def __init__(self):
+            self.posts = 0
+            self.errors = 0
+            self.successes = 0
+
+        def post(self, uri, *, data, **kwargs):
+            self.posts += 1
+            try:
+                json_data = json.loads(data.decode("utf-8"))
+                assert_serial_nos_ok([log["serial_no"] for log in json_data["logs"]])
+                self.successes += 1
+            except Exception:
+                self.errors += 1
+
+    with ExitStack() as exit_stack:
+        session = MockSession()
+        handler = HttpBatchRequestsHandler("localhost:61234", flush_interval=2.0)
+        handler.session = session
+        exit_stack.callback(handler.stop)
+
+        logging.basicConfig(force=True, handlers=[handler], level=0)
+
+        def log_func(end_time):
+            while time.time() < end_time:
+                logging.info("A" * random.randint(50, 4000))
+                logging.info("A" * random.randint(50, 4000))
+                logging.info("A" * random.randint(50, 4000))
+
+        end_time = time.time() + 20.0
+        threads = [
+            Thread(target=log_func, name="Log thread 1", args=(end_time,)),
+            Thread(target=log_func, name="Log thread 2", args=(end_time,)),
+            Thread(target=log_func, name="Log thread 3", args=(end_time,)),
+        ]
+
+        for t in threads:
+            t.start()
+
+        for t in threads:
+            t.join()
+
+        assert session.posts > 5
+        assert session.successes == session.posts
+        assert session.errors == 0
```

Comment:
Missing a test on retry logic (that attempts to try X times before dropping the buffer - need to simulate and test this flow somehow)